### PR TITLE
Add granim typings

### PR DESCRIPTION
**Pull request by Raiper34**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` should be present and it shouldn't have any additional or disabling of rules. Just content as `{ "extends": "dtslint/dt.json" }`. If for reason the some rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]`  and not for whole package so that the need for disabling can be reviewed.
- [x] `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Review by peterblazejewicz — COMMENTED** [review]
@Raiper34
this looks more like CommonJS module with UMD export
https://github.com/sarcadass/granim.js/blob/master/lib/Granim.js#L128
https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html#identifying-kinds-of-libraries

**Comment by elibarzilay** [comment]
@Raiper34: Ping re @peterblazejewicz's comment, and also, see the instructions on the tslint config...

File: types/granim/tslint.json [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    "space-within-parens": false,
+    "strict-export-declare-modifiers": false,
+    "trim-file": false,
+    "triple-equals": false,
+    "typedef-whitespace": false,
+    "unified-signatures": false,
+    "void-return": false,
+    "whitespace": false
+  }
+}
```
Comment (peterblazejewicz) [inline types/granim/tslint.json:80]: Please don't do this, you should have a very specific reasons to do so (and documented in PR). Follow the path, would be better for users of this DT (I'll post rewritten code to comply with that change):
```suggestion
{ "extends": "dtslint/dt.json" }
```

File: types/granim/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+type stretchMode = 'none' | 'stretch' | 'stretch-if-smaller' | 'stretch-if-bigger';
+
+export interface Image {
+  source: string;
+  position?: ['left' | 'center' | 'right', 'top' | 'center' | 'bottom'];
+  stretchMode?: [stretchMode, stretchMode];
+  blendingMode?: string;
+}
```
Comment (peterblazejewicz) [inline types/granim/index.d.ts:54]: This is a content amended to comply with default DT linter settings, no brainer really, minor changes (redundant `declare`, types for arrays, export all types, etc):
```suggestion
export class Granim {
    constructor(options: Options);
    play(): void;
    pause(): void;
    clear(): void;
    changeState(stateName: string): void;
    changeDirection(directionName: string): void;
    changeBlendingMode(blendingModeName: string): void;
    destroy(): void;
}

export interface OnGradientChangeObject {
    activeState: string;
    colorsFrom: [string, string];
    colorsTo: [string, string];
    isLooping: boolean;
}

export interface Options {
    element: string | HTMLCanvasElement;
    name?: string;
    elToSetClassOn?: string;
    direction?: 'diagonal' | 'left-right' | 'top-bottom' | 'radial' | 'custom';
    customDirection?: { x0: string; y0: string; x1: string; y1: string };
    isPausedWhenNotInView?: boolean;
    scrollDebounceThreshold?: number;
    stateTransitionSpeed?: number;
    defaultStateName?: string;
    states: { [stateName: string]: State };
    image?: Image;
    onStart?: () => void;
    onGradientChange?: (colorDetails: OnGradientChangeObject) => void;
    onEnd?: () => void;
}

export interface Gradient {
    color: string;
    pos: number;
}

export interface State {
    gradients: string[][] | Gradient[][];
    transitionSpeed?: number;
    loop?: boolean;
}

export type stretchMode = 'none' | 'stretch' | 'stretch-if-smaller' | 'stretch-if-bigger';

export interface Image {
    source: string;
    position?: ['left' | 'center' | 'right', 'top' | 'center' | 'bottom'];
    stretchMode?: [stretchMode, stretchMode];
    blendingMode?: string;
}
```

File: types/granim/granim-tests.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        ['#4776E6', '#8E54E9']
+      ],
+      transitionSpeed: 2000
+    },
+    "orange-state": {
+      gradients: [ ['#FF4E50', '#F9D423'] ],
+      loop: false
+    }
+  }
+});
```
Comment (peterblazejewicz) [inline types/granim/granim-tests.ts:185]:
```suggestion
import { Granim } from 'granim';

// Tests from https://sarcadass.github.io/granim.js/api-v2.0.0.html

let granimInstance: Granim = new Granim({
    element: '',
    name: 'granim',
    elToSetClassOn: 'body',
    direction: 'diagonal',
    isPausedWhenNotInView: false,
    scrollDebounceThreshold: 300,
    stateTransitionSpeed: 1000,
    image: {
        source: '../assets/img/bg-forest.jpg',
        position: ['center', 'bottom'],
        stretchMode: ['stretch', 'stretch-if-bigger'],
        blendingMode: 'multiply',
    },
    states: {
        'default-state': {
            gradients: [
                ['#834d9b', '#d04ed6'],
                ['#1CD8D2', '#93EDC7'],
            ],
            transitionSpeed: 5000,
            loop: true,
        },
        'dark-state': {
            gradients: [
                ['#757F9A', '#D7DDE8'],
                ['#5C258D', '#4389A2'],
            ],
            transitionSpeed: 5000,
            loop: true,
        },
    },
    onStart: () => {
        console.log('Granim: onStart');
    },
    onGradientChange: colorDetails => {
        console.log('Granim: onGradientChange, details: ');
        console.log(colorDetails);
    },
    onEnd: () => {
        console.log('Granim: onEnd');
    },
});

granimInstance = new Granim({
    element: '',
    name: 'granim',
    elToSetClassOn: 'body',
    direction: 'diagonal',
    isPausedWhenNotInView: false,
    scrollDebounceThreshold: 300,
    stateTransitionSpeed: 1000,
    states: {
        'default-state': {
            gradients: [
                ['#834d9b', '#d04ed6', '#1CD8D2', '#93EDC7'],
                ['#1CD8D2', '#93EDC7', '#757F9A', '#4389A2'],
            ],
            transitionSpeed: 5000,
            loop: true,
        },
        'dark-state': {
            gradients: [
                ['#757F9A', '#D7DDE8', '#1CD8D2', '#93EDC7'],
                ['#5C258D', '#4389A2', '#1CD8D2', '#93EDC7'],
            ],
            transitionSpeed: 5000,
            loop: true,
        },
    },
});

// Tests from https://sarcadass.github.io/granim.js/examples.html

granimInstance = new Granim({
    element: '#canvas-basic',
    direction: 'left-right',
    isPausedWhenNotInView: true,
    states: {
        'default-state': {
            gradients: [
                ['#ff9966', '#ff5e62'],
                ['#00F260', '#0575E6'],
                ['#e1eec3', '#f05053'],
            ],
        },
    },
});

granimInstance = new Granim({
    element: '#canvas-complex',
    direction: 'left-right',
    isPausedWhenNotInView: true,
    states: {
        'default-state': {
            gradients: [
                [
                    { color: '#833ab4', pos: 0.2 },
                    { color: '#fd1d1d', pos: 0.8 },
                    { color: '#38ef7d', pos: 1 },
                ],
                [
                    { color: '#40e0d0', pos: 0 },
                    { color: '#ff8c00', pos: 0.2 },
                    { color: '#ff0080', pos: 0.75 },
                ],
            ],
        },
    },
});

granimInstance = new Granim({
    element: '#canvas-image-blending',
    direction: 'top-bottom',
    isPausedWhenNotInView: true,
    image: {
        source: '../assets/img/bg-forest.jpg',
        blendingMode: 'multiply',
    },
    states: {
        'default-state': {
            gradients: [
                ['#29323c', '#485563'],
                ['#FF6B6B', '#556270'],
                ['#80d3fe', '#7ea0c4'],
                ['#f0ab51', '#eceba3'],
            ],
            transitionSpeed: 7000,
        },
    },
});

granimInstance = new Granim({
    element: '#logo-canvas',
    direction: 'left-right',
    states: {
        'default-state': {
            gradients: [
                ['#EB3349', '#F45C43'],
                ['#FF8008', '#FFC837'],
                ['#4CB8C4', '#3CD3AD'],
                ['#24C6DC', '#514A9D'],
                ['#FF512F', '#DD2476'],
                ['#DA22FF', '#9733EE'],
            ],
            transitionSpeed: 2000,
        },
    },
});

granimInstance = new Granim({
    element: '#canvas-interactive',
    name: 'interactive-gradient',
    elToSetClassOn: '.canvas-interactive-wrapper',
    direction: 'diagonal',
    isPausedWhenNotInView: true,
    stateTransitionSpeed: 500,
    states: {
        'default-state': {
            gradients: [
                ['#B3FFAB', '#12FFF7'],
                ['#ADD100', '#7B920A'],
                ['#1A2980', '#26D0CE'],
            ],
            transitionSpeed: 10000,
        },
        'violet-state': {
            gradients: [
                ['#9D50BB', '#6E48AA'],
                ['#4776E6', '#8E54E9'],
            ],
            transitionSpeed: 2000,
        },
        'orange-state': {
            gradients: [['#FF4E50', '#F9D423']],
            loop: false,
        },
    },
});
```

File: types/granim/granim-tests.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
+import {Granim} from "granim";
```
Comment (peterblazejewicz) [inline types/granim/granim-tests.ts:1]: For some reason I cannot post a suggestion here,
Pick a content from this gist:
https://gist.github.com/peterblazejewicz/f0414196c7e1eddd5580769daf8a5af3

**Review by peterblazejewicz — CHANGES_REQUESTED** [review]
@Raiper34
see my recent suggestions, with them the definition will at least can be used with code that uses modules, while still complying with DT settings for linting, etc.
Sorry for this taking so long, but the intent here is just to provide easy to use definition files, that not only can be easily consumed, but also contributed to with new changes.

**Comment by Raiper34** [comment]
> @Raiper34
> see my recent suggestions, with them the definition will at least can be used with code that uses modules, while still complying with DT settings for linting, etc.
> Sorry for this taking so long, but the intent here is just to provide easy to use definition files, that not only can be easily consumed, but also contributed to with new changes.

I change to your suggestions, but currently test gives me
`ERROR: 1:1  npm-naming  The declaration doesn't match the JavaScript module 'granim'. Reason:
The declaration should use 'export =' syntax because the JavaScript source uses 'module.exports =' syntax and 'module.exports' can be called or constructed.` so I tried to rewrite into UMD module, but I am a bit lost now... Because when I auto import my typing, it is imported like import * as Granim from 'granim'; but in docs it is typed, that this is wrong import strategy...

**Comment by peterblazejewicz** [comment]
@Raiper34  what's wrong with this one? It's your current code with minorish change to suport umd.
it works with global script (note the export as namespace ...), CJS and ES imports.
```ts
// Type definitions for granim 2.0
// Project: https://github.com/sarcadass/granim.js
// Definitions by: Raiper34 <https://github.com/Raiper34>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped

export as namespace Granim;

declare class Granim {
    constructor(options: Granim.Options);
    play(): void;
    pause(): void;
    clear(): void;
    changeState(stateName: string): void;
    changeDirection(directionName: string): void;
    changeBlendingMode(blendingModeName: string): void;
    destroy(): void;
}

declare namespace Granim {
    export interface OnGradientChangeObject {
        activeState: string;
        colorsFrom: [string, string];
        colorsTo: [string, string];
        isLooping: boolean;
    }
    export interface Options {
        element: string | HTMLCanvasElement;
        name?: string;
        elToSetClassOn?: string;
        direction?: 'diagonal' | 'left-right' | 'top-bottom' | 'radial' | 'custom';
        customDirection?: { x0: string; y0: string; x1: string; y1: string };
        isPausedWhenNotInView?: boolean;
        scrollDebounceThreshold?: number;
        stateTransitionSpeed?: number;
        defaultStateName?: string;
        states: { [stateName: string]: State };
        image?: Image;
        onStart?: () => void;
        onGradientChange?: (colorDetails: OnGradientChangeObject) => void;
        onEnd?: () => void;
    }
    export interface Gradient {
        color: string;
        pos: number;
    }
    export interface State {
        gradients: string[][] | Gradient[][];
        transitionSpeed?: number;
        loop?: boolean;
    }
    export type stretchMode = 'none' | 'stretch' | 'stretch-if-smaller' | 'stretch-if-bigger';
    export interface Image {
        source: string;
        position?: ['left' | 'center' | 'right', 'top' | 'center' | 'bottom'];
        stretchMode?: [stretchMode, stretchMode];
        blendingMode?: string;
    }
}

export = Granim;
```

If I would use your current implementation, I will receive at runtime:
```text
var granimInstance = new granim_1.Granim({
                     ^

TypeError: granim_1.Granim is not a constructor
```

File: types/granim/granim-tests.ts [hunk 5]
```diff
@@ -0,0 +1,183 @@
+import { Granim } from 'granim';
```
Comment (peterblazejewicz) [inline types/granim/granim-tests.ts:1]:
```suggestion
import Granim = require('granim');
```
This is for DT test only, users could use `esModuleInterop` to default to more common syntax

File: types/granim/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,64 @@
+// Type definitions for granim 2.0
+// Project: https://github.com/sarcadass/granim.js
+// Definitions by: Raiper34 <https://github.com/Raiper34>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export as namespace Granim;
+
+declare class Granim {
+    constructor(options: Options);
```
Comment (peterblazejewicz) [inline types/granim/index.d.ts:9]:
```suggestion
    constructor(options: Granim.Options);
```

File: types/granim/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    play(): void;
+    pause(): void;
+    clear(): void;
+    changeState(stateName: string): void;
+    changeDirection(directionName: string): void;
+    changeBlendingMode(blendingModeName: string): void;
+    destroy(): void;
+}
+declare namespace Granim {
+    export interface OnGradientChangeObject {
```
Comment (peterblazejewicz) [inline types/granim/index.d.ts:19]:
```suggestion
    interface OnGradientChangeObject {
```

File: types/granim/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        activeState: string;
+        colorsFrom: [string, string];
+        colorsTo: [string, string];
+        isLooping: boolean;
+    }
+
+    export interface Options {
```
Comment (peterblazejewicz) [inline types/granim/index.d.ts:26]:
```suggestion
    interface Options {
```

File: types/granim/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        stateTransitionSpeed?: number;
+        defaultStateName?: string;
+        states: { [stateName: string]: State };
+        image?: Image;
+        onStart?: () => void;
+        onGradientChange?: (colorDetails: OnGradientChangeObject) => void;
+        onEnd?: () => void;
+    }
+
+    export interface Gradient {
```
Comment (peterblazejewicz) [inline types/granim/index.d.ts:43]:
```suggestion
    interface Gradient {
```

File: types/granim/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        color: string;
+        pos: number;
+    }
+
+    export interface State {
```
Comment (peterblazejewicz) [inline types/granim/index.d.ts:48]:
```suggestion
    interface State {
```

File: types/granim/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        gradients: string[][] | Gradient[][];
+        transitionSpeed?: number;
+        loop?: boolean;
+    }
+
+    export type stretchMode = 'none' | 'stretch' | 'stretch-if-smaller' | 'stretch-if-bigger';
```
Comment (peterblazejewicz) [inline types/granim/index.d.ts:54]:
```suggestion
    type stretchMode = 'none' | 'stretch' | 'stretch-if-smaller' | 'stretch-if-bigger';
```

File: types/granim/index.d.ts [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+    export interface Image {
```
Comment (peterblazejewicz) [inline types/granim/index.d.ts:56]:
```suggestion
    interface Image {
```

**Review by peterblazejewicz — COMMENTED** [review]
@Raiper34
This is ready to go,
I've added suggestions that fix your CI tests (CommonJS import for DT, `export`s are not required for DT namespace export)

**Comment by Raiper34** [comment]
> @Raiper34 This is ready to go, I've added suggestions that fix your CI tests (CommonJS import for DT, `export`s are not required for DT namespace export)

I reopened this PR and all checks passed, you can see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56221

**Pull request closed without merging**